ruler/scheduler.go

```diff
@@ -0,0 +1,229 @@
+package ruler
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/prometheus/rules"
+	"github.com/weaveworks/scope/common/instrument"
+	"golang.org/x/net/context"
+
+	"github.com/weaveworks/cortex/util"
+)
+
+const (
+	skewCorrection = 1 * time.Minute
+	// Backoff for loading initial configuration set.
+	minBackoff = 100 * time.Millisecond
+	maxBackoff = 2 * time.Second
+)
+
+var (
+	queueLength = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "cortex",
+		Name:      "rules_queue_length",
+		Help:      "The length of the rules queue.",
+	})
+	blockedWorkers = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "cortex",
+		Name:      "blocked_workers",
+		Help:      "How many workers are waiting on an item to be ready.",
+	})
+	totalConfigs = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "cortex",
+		Name:      "configs",
+		Help:      "How many configs the scheduler knows about.",
+	})
+	configsRequestDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "cortex",
+		Name:      "configs_request_duration_seconds",
+		Help:      "Time spent requesting configs.",
+		Buckets:   prometheus.DefBuckets,
+	}, []string{"operation", "status_code"})
+)
+
+func init() {
+	prometheus.MustRegister(queueLength)
+	prometheus.MustRegister(blockedWorkers)
+	prometheus.MustRegister(configsRequestDuration)
+	prometheus.MustRegister(totalConfigs)
+}
+
+type workItem struct {
+	userID    string
+	rules     []rules.Rule
+	scheduled time.Time
+}
+
+// Key implements Op
+func (w workItem) Key() string {
+	return w.userID
+}
+
+// Priority implements Op
+func (w workItem) Priority() int64 {
+	return -w.scheduled.Unix()
+}
+
+// Defer returns a copy of this work item, rescheduled to a later time.
+func (w workItem) Defer(interval time.Duration) workItem {
+	return workItem{w.userID, w.rules, w.scheduled.Add(interval)}
+}
+
+type scheduler struct {
+	configsAPI         configsAPI // XXX: Maybe make this an interface ConfigSource or similar.
+	evaluationInterval time.Duration
+	q                  *util.PriorityQueue
+
+	// All the configurations that we have. Only used for instrumentation.
+	cfgs map[string]cortexConfig
+
+	pollInterval time.Duration
+	lastPolled   time.Time
+	pollMutex    sync.RWMutex
+
+	quit chan struct{}
+	wait sync.WaitGroup
+}
+
+// newScheduler makes a new scheduler.
+func newScheduler(configsAPI configsAPI, evaluationInterval, pollInterval time.Duration) scheduler {
+	return scheduler{
+		configsAPI:         configsAPI,
+		evaluationInterval: evaluationInterval,
+		pollInterval:       pollInterval,
+		q:                  util.NewPriorityQueue(),
+		cfgs:               map[string]cortexConfig{},
+	}
+}
+
+// Run polls the source of configurations for changes.
+func (s *scheduler) Run() {
+	log.Debugf("Scheduler started")
+	s.wait.Add(1)
+	defer s.wait.Done()
```

Comment:
In the worker you used a channel to signal done, but here a wait group.  I don't feel strongly either way. But you should be consistent, and generally we've gone with a channel when its one thing, and a wait group when >1.